Geometry/HcalCommonData/src/HcalDDDSimConstants.cc

```diff
@@ -0,0 +1,791 @@
+      if (temp1.ok) {
+	int units = unitPhi (subdet0, eta);
+	HcalCellType temp2(subdet, eta, phi, depth, temp1,
+			   shift, gain, nz, nmod, hsize, units);
+	if (subdet == HcalOuter) {
+	  if (eta == hpar->noff[4]) {
+	    std::vector<int> missPlus, missMinus;
+	    int kk = 7;
+	    for (int miss=0; miss<hpar->noff[5]; miss++) {
+	      missPlus.push_back(hpar->noff[kk]);
+	      kk++;
+	    }
+	    for (int miss=0; miss<hpar->noff[6]; miss++) {
+	      missMinus.push_back(hpar->noff[kk]);
+	      kk++;
+	    }
+	    temp2.setMissingPhi(missPlus, missMinus);
+	  }
+	}
+	cellTypes.push_back(temp2);
+      }
+    }
+  }
+  return cellTypes;
+}
+
+unsigned int HcalDDDSimConstants::numberOfCells(HcalSubdetector subdet) const{
+
+  unsigned int num = 0;
+  std::vector<HcalCellType> cellTypes = HcalCellTypes(subdet);
+  for (unsigned int i=0; i<cellTypes.size(); i++) {
+    num += (unsigned int)(cellTypes[i].nPhiBins());
+    if (cellTypes[i].nHalves() > 1) 
+      num += (unsigned int)(cellTypes[i].nPhiBins());
+    num -= (unsigned int)(cellTypes[i].nPhiMissingBins());
+  }
+#ifdef DebugLog
+  edm::LogInfo ("HCalGeom") << "HcalDDDSimConstants:numberOfCells " 
+			<< cellTypes.size()  << " " << num 
+			<< " for subdetector " << subdet;
+#endif
+  return num;
+}
+
+int HcalDDDSimConstants::phiNumber(int phi, int units) const {
+
+  int iphi_skip = phi;
+  if      (units==2) iphi_skip  = (phi-1)*2+1;
+  else if (units==4) iphi_skip  = (phi-1)*4-1;
+  if (iphi_skip < 0) iphi_skip += 72;
+  return iphi_skip;
+}
+
+void HcalDDDSimConstants::printTiles() const {
+ 
+  std::cout << "Tile Information for HB from " << etaMin.at(0) << " to " << etaMin.at(0) << "\n\n";
+  for (int eta=etaMin.at(0); eta<= etaMax.at(0); eta++) {
+    int dmax = 1;
+    if (depths[0][eta-1] < 17) dmax = 2;
+    for (int depth=1; depth<=dmax; depth++) 
+      printTileHB(eta, depth);
+  }
+
+  std::cout << "\nTile Information for HE from " << etaMin.at(1) << " to " << etaMin.at(1) << "\n\n";
+  for (int eta=etaMin[1]; eta<= etaMax[1]; eta++) {
+    int dmin=1, dmax=3;
+    if (eta == etaMin[1]) {
+      dmin = 3;
+    } else if (depths[0][eta-1] > 18) {
+      dmax = 1;
+    } else if (depths[1][eta-1] > 18) {
+      dmax = 2;
+    }
+    for (int depth=dmin; depth<=dmax; depth++)
+      printTileHE(eta, depth);
+  }
+}
+
+int HcalDDDSimConstants::unitPhi(int det, int etaR) const {
+
+  double dphi = (det == static_cast<int>(HcalForward)) ? hpar->phitable[etaR-etaMin[2]] : hpar->phibin[etaR-1];
+  return unitPhi(dphi);
+}
+
+int HcalDDDSimConstants::unitPhi(double dphi) const {
+
+  const double fiveDegInRad = 2*M_PI/72;
+  int units = int(dphi/fiveDegInRad+0.5);
+  return units;
+}
+
+void HcalDDDSimConstants::initialize( void ) {
+
+  nEta      = hpar->etaTable.size();
+  nR        = hpar->rTable.size();
+  nPhiF     = nR - 1;
+  etaMin    = hpar->etaMin;
+  etaMax    = hpar->etaMax;
+  etaMin[0] = 1;
+  etaMax[1] = nEta-1;
+  etaMax[2] = etaMin[2]+nR-2;
```

Comment:
@bsunanda - are these numbers different from what is stored in HcalParameters and if yes, why? They are calculated directly from what is stored in the record any way. Using etaMin and etaMax vectors directly from HcalParameters record is more efficient - no need to copy it, no need to cache it.